src/Raven.Server/Documents/PeriodicBackup/DirectUpload/DirectUploadStream.cs

```diff
@@ -0,0 +1,173 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Client.Documents.Operations.Backups;
+using Raven.Client.Util;
+using Raven.Server.Documents.PeriodicBackup.Retention;
+using Sparrow;
+using Size = Sparrow.Size;
+
+namespace Raven.Server.Documents.PeriodicBackup.DirectUpload;
+
+public abstract class DirectUploadStream<T> : Stream where T : IDirectUploader
+{
+    private readonly IMultiPartUploader _multiPartUploader;
+    private readonly CloudUploadStatus _cloudUploadStatus;
+    private readonly Action<string> _onProgress;
+    private readonly IDisposable _backupStatusIDisposable;
+
+    private long _position;
+    private MemoryStream _writeStream = new();
+    private MemoryStream _uploadStream = new();
+    private Task _uploadTask;
+    private bool _disposed;
+
+    protected T Client { get; }
+
+    protected abstract long MaxPartSizeInBytes { get; }
+
+    protected DirectUploadStream(Parameters parameters)
+    {
+        _cloudUploadStatus = parameters.CloudUploadStatus;
+        _cloudUploadStatus.Skipped = false;
+        _backupStatusIDisposable = _cloudUploadStatus.UpdateStats(parameters.IsFullBackup);
+        _cloudUploadStatus.UploadProgress.ChangeState(UploadState.PendingUpload);
+        _onProgress = parameters.OnProgress;
+
+        var progress = Progress.Get(_cloudUploadStatus.UploadProgress, parameters.OnProgress);
+        Client = parameters.ClientFactory.Invoke(progress);
+        _multiPartUploader = Client.GetUploader(parameters.Key, parameters.Metadata);
+        _multiPartUploader.Initialize();
+    }
+
+    public override void Flush()
+    {
+    }
```

Comment:
Do we want to throw here as well?
```suggestion
    {
        throw new NotSupportedException();
    }
```

Reply:
not needed (just nothing to do here).